hooks/source_links.py

```
@@ -0,0 +1,145 @@
+    resolved = posixpath.normpath(posixpath.join(src_dir, path))
+    if not resolved.startswith('..'):
+        return None  # resolves within docs/ -- a normal internal link
+    repo_path = _LEADING_UP.sub('', resolved)
+    kind = 'tree' if path.endswith('/') else 'blob'
+    url = '{}/{}/{}/{}'.format(repo_url, kind, _ref(), repo_path)
+    if sep:
+        url += '#' + frag
+    return url
+
+
+def _fence_opener(line):
```

Ruff (0.15.20)

[warning] Missing return type annotation for private function `_fence_opener`

(ANN202)